File: Contents/InitialConsonantSearch/InitialConsonantSearch/Main/IndicatorSettingContentController.m
//
//  IndicatorSettingContentController.m
//  IndicatorSettingsTest
//
//  Created by Kwan Hyun Son on 10/10/23.
//

#import <BaseKit/BaseKit.h>
#import "IndicatorSettingContentController.h"
#import "IndicatorSettingDetailController.h"
#import "IndicatorSettingCell.h"

@interface IndicatorSettingContentController () <UITableViewDelegate, UITableViewDataSource>
@property (nonatomic, strong) UITableViewDiffableDataSource <NSString *, MGROutlineItem <DTOIndicatorSetting *>*>*dataSource;
@property (nonatomic, strong) UIStoryboard *board; // lazy
//@property (nonatomic, strong) IndicatorSettingDetailController *detailController; // lazy
@end

@implementation IndicatorSettingContentController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.tableView.rowHeight = 44.0;
    self.tableView.estimatedRowHeight = 44.0;
    /// 디퍼블을 이용하느냐 마느냐.
    [self configureDataSource];
}

- (void)configureDataSource {
    self.tableView.dataSource = nil;
    
    __weak __typeof(self) weakSelf = self;
    _dataSource =
    [[UITableViewDiffableDataSource alloc] initWithTableView:self.tableView
                                                cellProvider:^UITableViewCell *(UITableView *tableView,
                                                                                NSIndexPath *indexPath,
                                                                                MGROutlineItem <DTOIndicatorSetting *>*outlineItem) {
        __strong __typeof(weakSelf) self = weakSelf;
        
        NSString *identifier = outlineItem.contentItem.identifier;
        if ([weakSelf.viewModel.currentMainCategory isEqualToString:IndiSetMainCategoryFavorites]) {
            if ([identifier isEqualToString:IndicatorSettingCellIDSubMinus] == YES) {
                identifier = IndicatorSettingCellIDFavoriteSub;
            } else if ([identifier isEqualToString:IndicatorSettingCellIDSubPlus] == YES ||
                       [identifier isEqualToString:IndicatorSettingCellIDSubNormal] == YES ||
                       [identifier isEqualToString:IndicatorSettingCellIDRegion] == YES) {
                identifier = IndicatorSettingCellIDFavorite;
            }
        }
        
        IndicatorSettingCell *cell =
        [tableView dequeueReusableCellWithIdentifier:identifier
                                        forIndexPath:indexPath];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        if ([cell isKindOfClass:[IndicatorSettingCell class]] == NO) {
            NSAssert(FALSE, @"Could not create new cell");
        }
        
        cell.data = outlineItem.contentItem;
        return cell;
    }];

    self.dataSource.defaultRowAnimation = UITableViewRowAnimationFade;
    [self.dataSource applySnapshot:self.viewModel.snapshotForCurrentState animatingDifferences:NO];
}

#pragma mark - 세터 & 게터

- (UIStoryboard *)board {
    if (_board == nil) {
        _board = [UIStoryboard storyboardWithName:@"Main" bundle:[NSBundle mainBundle]];
    }
    return _board;
}

//- (IndicatorSettingDetailController *)detailController {
//    if (_detailController == nil) {
//        IndicatorSettingDetailController *vc = [self.board instantiateViewControllerWithIdentifier:@"IndicatorSettingDetailController"];
//        _detailController = vc;
//        ////_detailController.viewModel = self.viewModel;
//    }
//    return _detailController;
//}

#pragma mark - Actions

- (IBAction)selectionButtonClicked:(UIButton *)sender {
    NSIndexPath *indexPath = [self.tableView mgrIndexPathOfCellWhereViewExists:sender];
    MGROutlineItem <DTOIndicatorSetting *>*outlineItem = [self.dataSource itemIdentifierForIndexPath:indexPath];
    IndicatorSettingCell *cell = (IndicatorSettingCell *)[self.tableView cellForRowAtIndexPath:indexPath];
    outlineItem.contentItem.selected = !(outlineItem.contentItem.selected);
    cell.data = cell.data;
    MGROutlineItem <DTOIndicatorSetting *>*progenitor = outlineItem.progenitor;
    NSIndexPath *progenitorPath = [self.dataSource indexPathForItemIdentifier:progenitor];
    IndicatorSettingCell *progenitorCell = [self.tableView cellForRowAtIndexPath:progenitorPath];
    progenitorCell.data = progenitor.contentItem;
}

- (IBAction)favButtonClicked:(UIButton *)sender {
    MGROutlineItem <DTOIndicatorSetting *>*outlineItem = [self outlineDataForSender:sender];
    DTOIndicatorSetting *data = [self dataForSender:sender];
    IndicatorSettingCell *cell = [self cellForSender:sender];
    data.favorite = !data.favorite;
    cell.data = data;
    
    __weak __typeof(self) weakSelf = self;
    [self.viewModel updateFavoriteItem:outlineItem completion:^{
        if ([weakSelf.viewModel.currentMainCategory isEqualToString:IndiSetMainCategoryFavorites]) {
            [weakSelf.dataSource applySnapshot:weakSelf.viewModel.snapshotForCurrentState animatingDifferences:YES];
        }
    }];
}

- (IBAction)plusButtonClicked:(UIButton *)sender {
    NSIndexPath *indexPath = [self.tableView mgrIndexPathOfCellWhereViewExists:sender];
    MGROutlineItem <DTOIndicatorSetting *>*outlineItem = [self.dataSource itemIdentifierForIndexPath:indexPath];
    [self.viewModel addChildItemForItem:outlineItem];
    [self.dataSource applySnapshot:self.viewModel.snapshotForCurrentState
              animatingDifferences:YES
                        completion:^{}];
}

- (IBAction)minusButtonClicked:(UIButton *)sender {
    NSIndexPath *indexPath = [self.tableView mgrIndexPathOfCellWhereViewExists:sender];
    MGROutlineItem <DTOIndicatorSetting *>*outlineItem = [self.dataSource itemIdentifierForIndexPath:indexPath];
    MGROutlineItem <DTOIndicatorSetting *>*progenitor = outlineItem.progenitor;
    [self.viewModel deleteItem:outlineItem];
    NSIndexPath *progenitorPath = [self.dataSource indexPathForItemIdentifier:progenitor];
    IndicatorSettingCell *progenitorCell = [self.tableView cellForRowAtIndexPath:progenitorPath];
    progenitorCell.data = progenitor.contentItem;
    [self.dataSource applySnapshot:self.viewModel.snapshotForCurrentState
              animatingDifferences:YES
                        completion:^{}];
}

- (IBAction)settingButtonClicked:(UIButton *)sender {
    
    NSString *subString = IndiSetMainCategoryIndicators;
    if ([self.viewModel.currentMainCategory isEqualToString:IndiSetMainCategoryIndicators]) {
        // 그대로
    } else if ([self.viewModel.currentMainCategory isEqualToString:IndiSetMainCategorySignals]) {
        subString = IndiSetMainCategorySignals;
    } else if ([self.viewModel.currentMainCategory isEqualToString:IndiSetMainCategoryPatterns]) {
        subString = IndiSetMainCategoryPatterns;
    } else if ([self.viewModel.currentMainCategory isEqualToString:IndiSetMainCategoryRanges]) {
        subString = IndiSetMainCategoryRanges;
    } else if ([self.viewModel.currentMainCategory isEqualToString:IndiSetMainCategoryFill]) {
        subString = IndiSetMainCategoryFill;
    } else if ([self.viewModel.currentMainCategory isEqualToString:IndiSetMainCategoryFavorites]) {
        subString = IndiSetMainCategoryFavorites;
    } else {
        NSCAssert(FALSE, @"예상치 못한 값이 들어왔다. 수정하라");
    }
    NSString *title = [NSString stringWithFormat:@"지표설정/%@", subString];
    self.navigationController.topViewController.navigationItem.backButtonTitle = title;
    DTOIndicatorSetting *contentItem = [self dataForSender:sender];
    UIViewController *vc = [self settingViewControllerForIndicatorSetting:contentItem];
    [self.navigationController pushViewController:vc animated:YES];
}

- (void)reloadData:(void(^)(void))completion {
    [self.dataSource applySnapshotUsingReloadData:self.viewModel.snapshotForCurrentState completion:completion];
}

- (void)scrollToRowAtItem:(MGROutlineItem <DTOIndicatorSetting *>*)item
         atScrollPosition:(UITableViewScrollPosition)scrollPosition
                 animated:(BOOL)animated {
    NSInteger index = [self.dataSource.snapshot indexOfItemIdentifier:item];
    if (index != NSNotFound) {
        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:index inSection:0];
        [self.tableView scrollToRowAtIndexPath:indexPath
                              atScrollPosition:scrollPosition
                                      animated:animated];
    }
}


- (UIViewController *)settingViewControllerForIndicatorSetting:(DTOIndicatorSetting *)indicatorSetting {
    IndicatorSettingDetailController *vc = [self.board instantiateViewControllerWithIdentifier:@"IndicatorSettingDetailController"];
    
    vc.viewModel = [self.viewModel detailViewModelForIndicatorSetting:indicatorSetting];
    return vc;
}

#pragma mark - <UITableViewDataSource>
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 3;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    return nil;
}

#pragma mark - <UITableViewDelegate>

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    MGROutlineItem <DTOIndicatorSetting *>*outlineItem = [self.dataSource itemIdentifierForIndexPath:indexPath];
    IndicatorSettingCell *cell = (IndicatorSettingCell *)[tableView cellForRowAtIndexPath:indexPath];
    if ([outlineItem.contentItem.identifier isEqualToString:IndicatorSettingCellIDSection] == YES &&
        outlineItem.hasSubitem == YES) {
        outlineItem.expanded = !outlineItem.isExpanded;
        
        [UIViewPropertyAnimator runningPropertyAnimatorWithDuration:0.3
                                                              delay:0.0
                                                            options:kNilOptions
                                                         animations:^{ cell.data = cell.data; }
                                                         completion:^(UIViewAnimatingPosition finalPosition) {}];
        [self.dataSource applySnapshot:self.viewModel.snapshotForCurrentState
                  animatingDifferences:YES
                            completion:^{}];
    }
}


#pragma mark - Helper
- (IndicatorSettingCell *)cellForSender:(__kindof UIView *)sender {
    NSIndexPath *indexPath = [self.tableView mgrIndexPathOfCellWhereViewExists:sender];
    return [self.tableView cellForRowAtIndexPath:indexPath];
}

- (DTOIndicatorSetting *)dataForSender:(__kindof UIView *)sender {
    return [self outlineDataForSender:sender].contentItem;
}

- (MGROutlineItem <DTOIndicatorSetting *>*)outlineDataForSender:(__kindof UIView *)sender {
    NSIndexPath *indexPath = [self.tableView mgrIndexPathOfCellWhereViewExists:sender];
    return [self.dataSource itemIdentifierForIndexPath:indexPath];
}

@end

File: Contents/InitialConsonantSearch/InitialConsonantSearch_Swift/Main/IndicatorSettingContentController.swift
//
//  IndicatorSettingContentController.swift
//  IndicatorSettingsTest_Swift
//
//  Created by Kwan Hyun Son on 10/12/23.
//

import UIKit
import BaseKit
import IosKit

final class IndicatorSettingContentController: UIViewController {

    // MARK: - Property
    
    @IBOutlet weak var tableView: UITableView!
    var dataSource: UITableViewDiffableDataSource<String, SKHOutlineItem<DTOIndicatorSetting>>?
    var viewModel: IndicatorSettingViewModel?
    lazy var board: UIStoryboard = {
        return UIStoryboard(name: "Main", bundle: Bundle.main)
    }()
    lazy var detailController: IndicatorSettingDetailController = {
        let controller = board.instantiateViewController(withIdentifier: "IndicatorSettingDetailController") as! IndicatorSettingDetailController
//        controller.viewModel = self.viewModel
        return controller
    }()


    // MARK: - Override
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.rowHeight = 44.0
        tableView.estimatedRowHeight = 44.0
        // tableView.contentInsetAdjustmentBehavior = .never
        // tableView.contentInset = UIEdgeInsets(top: 0.0, left: 0.0, bottom: 1000.0, right: 0.0)
        configureDataSource()
    }
    
    // MARK: - 생성 & 소멸
    
    private func configureDataSource() {
        tableView.dataSource = nil
        
        dataSource = UITableViewDiffableDataSource<String, SKHOutlineItem<DTOIndicatorSetting>>(
            tableView: tableView,
            cellProvider: { (tableView: UITableView, indexPath: IndexPath, outlineItem: SKHOutlineItem<DTOIndicatorSetting>) -> IndicatorSettingCell? in
                
                guard var identifier = outlineItem.contentItem?.identifier
                else {
                    return nil
                }
                if self.viewModel?.currentMainCategory == .favorites {
                    if identifier == IndicatorSettingCellID.subMinus.rawValue {
                        identifier = IndicatorSettingCellID.favoriteSub.rawValue
                    } else if identifier == IndicatorSettingCellID.subPlus.rawValue ||
                              identifier == IndicatorSettingCellID.subNormal.rawValue ||
                              identifier == IndicatorSettingCellID.region.rawValue {
                        identifier = IndicatorSettingCellID.favorite.rawValue
                    } else {
                        assertionFailure("예상치 못한 identifier이다.")
                    }
                }
                
                guard let cell = tableView.dequeueReusableCell(withIdentifier: identifier, for: indexPath) as? IndicatorSettingCell
                else {
                    assertionFailure("Could not create new cell")
                    return nil
                }
                cell.selectionStyle = .none
                cell.data = outlineItem.contentItem
                return cell
            }
        )
        
        guard let dataSource = dataSource,
              let viewModel = viewModel
        else {
            return
        }
        dataSource.defaultRowAnimation = .fade
        dataSource.apply(viewModel.snapshotForCurrentState(), animatingDifferences: false)
    }
    
    // MARK: - Actions
    
    @IBAction func selectionButtonClicked(_ sender: UIButton) {
        guard let indexPath = tableView.skhIndexPathOfCellWhereViewExists(view: sender),
              let outlineItem = dataSource?.itemIdentifier(for: indexPath)
        else {
            return
        }

        outlineItem.contentItem?.isSelected.toggle()
        
        if let cell = tableView.cellForRow(at: indexPath) as? IndicatorSettingCell,
           let data = cell.data {
            cell.data = data
        }
        
        if let progenitor = outlineItem.progenitor,
           let progenitorPath = dataSource?.indexPath(for: progenitor),
           let progenitorCell = tableView.cellForRow(at: progenitorPath) as? IndicatorSettingCell {
            progenitorCell.data = progenitor.contentItem
        }
    }
    
    @IBAction func favButtonClicked(_ sender: UIButton) {
        guard let outlineItem = outlineDataForSender(sender),
              let data = outlineItem.contentItem,
              let viewModel = viewModel
        else {
            return
        }

        data.isFavorite.toggle()
        if let cell = cellForSender(sender) {
            cell.data = data
        }
        
        if #available(iOS 13, *) {
            var completionClosure : (() -> Void)?
            if viewModel.currentMainCategory == .favorites {
                completionClosure = { [weak self, weak viewModel] () -> Void in
                    if let strongviewModel = viewModel {
                        self?.dataSource?.apply(
                            strongviewModel.snapshotForCurrentState(),
                            animatingDifferences: true,
                            completion: nil)
                    }
                }
            }
            viewModel.updateFavoriteItem(outlineItem, completion: completionClosure)
        }
    }
    
    @IBAction func plusButtonClicked(_ sender: UIButton) {
        guard let dataSource = dataSource,
              let viewModel = viewModel,
              let indexPath = tableView.skhIndexPathOfCellWhereViewExists(view: sender),
              let outlineItem = dataSource.itemIdentifier(for: indexPath)
        else {
            return
        }
        viewModel.addChildItem(for: outlineItem)
        dataSource.apply(viewModel.snapshotForCurrentState(), animatingDifferences: true, completion: nil)
    }
    
    @IBAction func minusButtonClicked(_ sender: UIButton) {
        guard let dataSource = dataSource,
              let viewModel = viewModel,
              let indexPath = tableView.skhIndexPathOfCellWhereViewExists(view: sender),
              let outlineItem = dataSource.itemIdentifier(for: indexPath)
        else {
            return
        }
        let progenitor = outlineItem.progenitor // 반드시 먼저 잡아야한다
        viewModel.deleteItem(outlineItem)
        if let progenitor = progenitor,
           let progenitorPath = dataSource.indexPath(for: progenitor),
           let progenitorCell = tableView.cellForRow(at: progenitorPath) as? IndicatorSettingCell {
            progenitorCell.data = progenitor.contentItem
        }
        dataSource.apply(viewModel.snapshotForCurrentState(), animatingDifferences: true, completion: nil)
    }
    
    @IBAction func settingButtonClicked(_ sender: UIButton) {
        if let navigationController = navigationController {
            var subString = IndiSetMainCategory.indicators.rawValue
            if viewModel?.currentMainCategory == .indicators {
                // 그대로
            } else if viewModel?.currentMainCategory == .signals {
                subString = IndiSetMainCategory.signals.rawValue
            } else if viewModel?.currentMainCategory == .patterns {
                subString = IndiSetMainCategory.patterns.rawValue
            } else if viewModel?.currentMainCategory == .ranges {
                subString = IndiSetMainCategory.ranges.rawValue
            } else if viewModel?.currentMainCategory == .fill {
                subString = IndiSetMainCategory.fill.rawValue
            } else if viewModel?.currentMainCategory == .favorites {
                subString = IndiSetMainCategory.favorites.rawValue
            } else {
                assertionFailure("예상치 못한 값이 들어왔다. 수정하라")
            }
            
            navigationController.topViewController?.navigationItem.backButtonTitle = "지표설정/" + subString
            navigationController.pushViewController(detailController, animated: true)
        }
    }
    
    func reloadData(completion: (() -> Void)? = nil) {
        if #available(iOS 13.0, *) {
            guard let dataSource = dataSource,
                  let viewModel = viewModel else {
                return
            }
            if #available(iOS 15.0, *) {
                dataSource.applySnapshotUsingReloadData(viewModel.snapshotForCurrentState()) {
                    completion?()
                }
            } else {
                dataSource.apply(viewModel.snapshotForCurrentState(), animatingDifferences: false) {
                    completion?()
                }
            }
        }
    }
    
    func scrollToRow(
        at item: SKHOutlineItem<DTOIndicatorSetting>,
        at scrollPosition: UITableView.ScrollPosition,
        animated: Bool) {
            if let index = dataSource?.snapshot().indexOfItem(item) {
                let indexPath = IndexPath.init(row: index, section: 0)
                tableView.scrollToRow(at: indexPath, at: scrollPosition, animated: animated)
            }
    }
    
    // MARK: - Helper

    private func cellForSender(_ sender: UIView) -> IndicatorSettingCell? {
        if let indexPath = tableView.skhIndexPathOfCellWhereViewExists(view: sender) {
            return tableView.cellForRow(at: indexPath) as? IndicatorSettingCell
        }
        return nil
    }
    
    private func dataForSender(_ sender: UIView) -> DTOIndicatorSetting? {
        return outlineDataForSender(sender)?.contentItem
    }
    
    private func outlineDataForSender(_ sender: UIView) -> SKHOutlineItem<DTOIndicatorSetting>? {
        if #available(iOS 13.0, *) {
            if let indexPath = tableView.skhIndexPathOfCellWhereViewExists(view: sender),
               let outlineItem = dataSource?.itemIdentifier(for: indexPath) {
                return outlineItem
            }
        }
        return nil
    }
}

extension IndicatorSettingContentController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard let dataSource = dataSource,
              let outlineItem = dataSource.itemIdentifier(for: indexPath),
              let cell = tableView.cellForRow(at: indexPath) as? IndicatorSettingCell else {
            return
        }
            
        if outlineItem.contentItem?.identifier == IndicatorSettingCellID.section.rawValue &&
           outlineItem.hasSubitem {
            outlineItem.isExpanded.toggle()
            UIViewPropertyAnimator.runningPropertyAnimator(
                withDuration: 0.3,
                delay: 0.0,
                options: [],
                animations: {
                    cell.data = cell.data
                },
                completion: { _ in }
            )
            if let viewModel = self.viewModel {
                self.dataSource?.apply(
                    viewModel.snapshotForCurrentState(),
                    animatingDifferences: true,
                    completion: {}
                )
            }
        }
    }
}

extension IndicatorSettingContentController: UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        3
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "ToolSettingCell", for: indexPath)
        cell.selectionStyle = .none
        return cell
    }
}
